pallets/confidential/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(box_syntax)]
+
+use polymesh_common_utilities::{identity::Trait as IdentityTrait, Context};
+use polymesh_primitives::{IdentityId, Ticker};
+use polymesh_primitives_derive::{SliceU8StrongTyped, VecU8StrongTyped};
+
+use pallet_identity as identity;
+
+use bulletproofs::RangeProof;
+use cryptography::asset_proofs::range_proof::{
+    prove_within_range, verify_within_range, InRangeProof,
+};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+
+use codec::{Decode, Encode};
+use frame_support::{
+    debug, decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult,
+    weights::SimpleDispatchInfo,
+};
+use frame_system::{self as system, ensure_signed};
+use sp_std::prelude::*;
+
+pub mod rng;
+pub use rng::native_rng;
+
+#[derive(Encode, Decode, Clone, Debug, Default, PartialEq, Eq, SliceU8StrongTyped)]
+pub struct RangeProofInitialMessageWrapper(pub [u8; 32]);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped)]
+pub struct RangeProofFinalResponseWrapper(pub Vec<u8>);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct TickerRangeProof {
+    // pub proof: RangeProofWrapper,
+    pub initial_message: RangeProofInitialMessageWrapper,
+    // pub committed_value: CompressedRistrettoWrapper,
+    pub final_response: RangeProofFinalResponseWrapper,
+    pub max_two_exp: u32,
+}
+
+pub trait Trait: frame_system::Trait + IdentityTrait {
+    type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+type Identity<T> = identity::Module<T>;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct ProverTickerKey {
+    prover: IdentityId,
+    ticker: Ticker,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Confidential {
+        /// Number of investor per asset.
+        pub RangeProofs get(fn range_proof): double_map hasher(twox_64_concat) IdentityId, hasher(blake2_128_concat) ProverTickerKey => Option<TickerRangeProof>;
+
+        pub RangeProofVerifications get(fn range_proof_verification): double_map hasher(blake2_128_concat) (IdentityId, Ticker), hasher(twox_64_concat) IdentityId => bool;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        #[weight = SimpleDispatchInfo::FixedNormal(400_000)]
+        pub fn add_range_proof(origin,
+            target_id: IdentityId,
+            ticker: Ticker,
+            secret_value: u64,
```

Comment:
What exactly is the "secret_value"? It will become public as soon the transaction is broadcast (even before it is mined). If this is something others can "steal", we can't take it as an input like this.

Reply:
This module is just for testing the integration in native and wasm. Its functionality will be replaced soon, and it exists only to verify that we are able to call `cryptography` functions from core.